src/table-utils/table-desc.ts

```diff
@@ -0,0 +1,190 @@
+import isNumber from 'is-number';
+import type {Node} from 'prosemirror-model';
+
+import {isTableBodyNode, isTableNode} from './utils';
+
+type TableCellRealDesc = {
+    type: 'real';
+    offset: number;
+    node: Node;
+    colspan?: number;
+    rowspan?: number;
+};
+type TableCellVirtualDesc = {
+    type: 'virtual';
+    /** Index of real cell, that prodeced this virtual cell */
+    colspan?: [number, number];
+    /** Index of real cell, that prodeced this virtual cell */
+    rowspan?: [number, number];
+};
+
+type Pos = {from: number; to: number};
+export type CellPos = ({type: 'real'} & Pos) | {type: 'virtual'; closestPos: number};
+
+type TableCellDesc = TableCellRealDesc | TableCellVirtualDesc;
+
+type TableRowDesc = {
+    node: Node;
+    cells: TableCellDesc[];
+    offset: number;
+    colspan?: boolean;
+    rowspan?: boolean;
+};
+
+export class TableDesc {
+    private static __cache = new WeakMap<Node, TableDesc>();
+
+    static create(table: Node) {
+        if (this.__cache.has(table)) return this.__cache.get(table)!;
+
+        if (!isTableNode(table)) return null;
+        const tbody = table.lastChild;
+        if (!tbody || !isTableBodyNode(tbody)) return null;
+
+        // table –> tbody -> ...
+        const baseOffset = 2;
+
+        const rows: TableRowDesc[] = [];
+        tbody.forEach((trow, offset) => {
+            rows.push({node: trow, offset, cells: []});
+        });
+
+        tbody.forEach((trow, _1, trowIndex) => {
```

Comment:
Perhaps the problem can be solved in `O(N)` - if you go through the cells and put metadata about them in a hash map.

At the same time, I assume that the keys of the HashMap will contain only those cell numbers that have colspan or rowspan attributes, so most often, the size of this object in memory will not be maximal.